lib/engine/auto_router.rb

```diff
@@ -88,81 +133,328 @@ def compute(corporation, **opts)
 
       train_routes.each do |train, routes|
         train_routes[train] = routes.sort_by(&:revenue).reverse.take(route_limit)
       end
 
-      train_routes = train_routes.values.sort_by(&:size)
+      sorted_routes = train_routes.map { |_train, routes| routes }
 
-      combos = [[]]
-      possibilities = []
+      limit = sorted_routes.map(&:size).reduce(&:*)
+      puts "Finding route combos of best #{train_routes.map { |k, v| k.name + ':' + v.size.to_s }.join(', ')} "\
+           "routes with depth #{limit}"
 
-      limit = train_routes.map(&:size).reduce(&:*)
-      puts "Finding route combos with depth #{limit}"
-      counter = 0
-      now = Time.now
+      possibilities = js_evaluate_combos(sorted_routes, route_timeout)
 
-      train_routes.each do |routes|
-        combos = routes.flat_map do |route|
-          combos.map do |combo|
-            combo += [route]
-            route.routes = combo
-            route.clear_cache!(only_routes: true)
-            counter += 1
-            if (counter % 1000).zero?
-              puts "#{counter} / #{limit}"
-              raise if Time.now - now > route_timeout
-            end
+      # final sanity check on best combos: recompute each route.revenue in case it needs to reject a combo
+      bad_route = nil
+      max_routes = possibilities.max_by do |routes|
+        routes.each do |route|
+          route.clear_cache!(only_routes: true)
+          route.routes = routes
+          bad_route = route
+          route.revenue
+        end
+        @game.routes_revenue(routes)
+      rescue GameError => e
+        # report error but still include combo with errored route in the result set
+        puts " Sanity check error, likely an auto_router bug: #{e}"
+        print_bad_route(routes, bad_route)
+        routes
+      end || []
+
+      max_routes.each { |route| route.routes = max_routes }
+    end
 
-            route.revenue
-            possibilities << combo
-            combo
-          rescue GameError # rubocop:disable Lint/SuppressedException
+    def print_bad_route(routes, bad_route)
+      puts "   route: #{bad_route.connection_hexes.map { |hex| hex }.join('-')}"
+      puts '   combo routes by hex:'
+      routes.each do |route|
+        puts "          #{route.connection_hexes.map { |hex| hex }.join('-')}"
+      end
+    end
+
+    # inputs:
+    #   connection is a route's connection_data
+    #   hexside_bits is a map of hexside_id to bit number
+    #   bit_to_bitmask is a helper table of bit 0-31 to integer bitmask (essentially 1 << bit)
+    # returns:
+    #   the bitfield (array of ints) representing all hexsides in the connection path
+    # updates:
+    #   new hexsides are added to hexside_bits
+    def bitfield_from_connection(connection, hexside_bits, bit_to_bitmask)
+      bitfield = [0]
+      connection.each do |conn|
+        paths = conn[:chain][:paths]
+        index = 1
+        stop = paths.size
+        node2 = paths[0]
+        while index < stop
```

Comment:
why not just do

paths.size.times do |i|
... 

i doubt this hand rolled loop is much faster?